vnext/ReactUWP/Views/Image/ReactImage.cpp

```diff
@@ -53,66 +53,207 @@ void ReactImage::OnLoadEnd(winrt::event_token const &token) noexcept {
-                winrt::LoadedImageSourceLoadCompletedEventArgs const &args) {
+              if (fireLoadEndEvent) {
+                strong_this->m_onLoadEndEvent(*strong_this, succeeded);
+              }
+
+              strong_this->m_sizeChangedRevoker.revoke();
+            }
+          });
+    } else {
+      winrt::ImageBrush imageBrush{strong_this->Background().try_as<winrt::ImageBrush>()};
+      bool createImageBrush{!imageBrush};
+      if (createImageBrush) {
+        imageBrush = winrt::ImageBrush{};
+
+        // ImageOpened and ImageFailed are mutually exclusive. One event of the other will
+        // always fire whenever an ImageBrush has the ImageSource value set or reset.
+        strong_this->m_imageBrushOpenedRevoker = imageBrush.ImageOpened(
+            winrt::auto_revoke, [weak_this, imageBrush, fireLoadEndEvent](const auto &, const auto &) {
               if (auto strong_this{weak_this.get()}) {
-                bool succeeded{false};
-                if (args.Status() == winrt::LoadedImageSourceLoadStatus::Success) {
-                  strong_this->m_brush->Source(surface);
-                  succeeded = true;
+                const auto bitmap{imageBrush.ImageSource().as<winrt::BitmapImage>()};
+                strong_this->m_imageSource.height = bitmap.PixelHeight();
+                strong_this->m_imageSource.width = bitmap.PixelWidth();
+
+                imageBrush.Stretch(strong_this->ResizeModeToStretch(strong_this->m_resizeMode));
+
+                if (fireLoadEndEvent) {
+                  strong_this->m_onLoadEndEvent(*strong_this, true);
                 }
-                strong_this->m_onLoadEndEvent(*strong_this, succeeded);
+              }
+            });
+
+        strong_this->m_imageBrushFailedRevoker =
+            imageBrush.ImageFailed(winrt::auto_revoke, [weak_this, fireLoadEndEvent](const auto &, const auto &) {
+              const auto strong_this{weak_this.get()};
+              if (strong_this && fireLoadEndEvent) {
+                strong_this->m_onLoadEndEvent(*strong_this, false);
               }
             });
       }
+
+      winrt::BitmapImage bitmapImage{imageBrush.ImageSource().try_as<winrt::BitmapImage>()};
+
+      if (!bitmapImage) {
+        bitmapImage = winrt::BitmapImage{};
+
+        strong_this->m_bitmapImageOpened = bitmapImage.ImageOpened(
+            winrt::auto_revoke, [imageBrush](const auto &, const auto &) { imageBrush.Opacity(1); });
+
+        imageBrush.ImageSource(bitmapImage);
+      }
+
+      if (createImageBrush) {
+        strong_this->Background(imageBrush);
+      }
+
+      if (fromStream) {
+        co_await bitmapImage.SetSourceAsync(memoryStream);
+      } else {
+        bitmapImage.UriSource(uri);
+
+        // TODO: When we change the source of a BitmapImage, we're getting a flicker of the old image
+        // being resized to the size of the new image. This is a temporary workaround.
+        imageBrush.Opacity(0);
```

Comment:
is the workaround expected to eliminate the flickering? I see the flickering you're talking about using these changes